### DISCUSSION TITLE
basic-extra-meal/int-cache

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#254] [Announcements]
# basic-extra-meal/int-cache

本文详细介绍了Java基本数据类型缓存池，包括其工作原理、应用场景以及如何使用缓存池提高内存利用效率。通过本文，您将了解到Java基本数据类型缓存池的优化策略和实践，掌握如何在实际开发中合理使用缓存池，提高程序性能。

https://javabetter.cn/basic-extra-meal/int-cache.html

**Comment by XieShangxu** [comment]
在JDK18里面，
```
Integer x = new Integer(18);
Integer y = new Integer(18);
System.out.println(x == y);
```
已经是true了

**Reply by XieShangxu** [reply]
搞错了。。

**Reply by KunCodeAscension** [reply]
JDK9以后这种方法创建Integer对象的API都被废除了，你怎么创建的

**Comment by Git-jimmyLee-hub** [comment]
当我们通过 Integer.valueOf() 方法获取整数对象时，会先检查该整数是否在 IntegerCache 中，如果在，则返回缓存中的对象，否则创建一个新的对象并缓存起来。这句话应该是会返回新对象，但不会缓存起来

**Comment by lucky-Luor** [comment]
当我们通过 Integer.valueOf() 方法获取整数对象时，会先检查该整数是否在 IntegerCache 中，如果在，则返回缓存中的对象，否则创建一个新的对象并缓存起来。 这句话确实是错误的,IntegerCache 是静态初始化的缓存池，仅在类加载时一次性创建 -128 ~ 127 范围内的对象并缓存，超出该范围的整数调用 valueOf() 时，只会创建新对象，不会加入缓存。
// 127在缓存范围内，返回缓存对象（同一地址）
Integer a = Integer.valueOf(127);
Integer b = Integer.valueOf(127);
System.out.println(a == b); // true

// 128超出默认缓存范围，创建新对象（不同地址）
Integer c = Integer.valueOf(128);
Integer d = Integer.valueOf(128);
System.out.println(c == d); // false
